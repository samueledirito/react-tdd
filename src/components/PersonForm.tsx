import React from "react";
import { Formik, Form } from "formik";
import Button from "@material-ui/core/Button";
import * as yup from "yup";
import TextField from "formik/TextField";

const validationSchema = yup.object().shape({
  personName: yup.string().required(),
});

const PersonForm: React.FC<{ person?: any }> = ({
  person = { personName: "" },
}) => {
  return (
    <Formik
      validationSchema={validationSchema}
      onSubmit={console.log}
      initialValues={person}
    >
      {({ dirty, isValid }) => (
        <Form>
          <TextField
            label="Name"
            name="personName"
            data-testid="personName"
            placeholder="Name"
          />
          <Button type="submit" disabled={!dirty || !isValid}>
            Save
          </Button>
        </Form>
      )}
    </Formik>
  );
};

export default PersonForm;
