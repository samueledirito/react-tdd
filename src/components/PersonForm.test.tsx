import React from "react";
import { render, fireEvent } from "@testing-library/react";

import PersonForm from "components/PersonForm";

describe("PersonForm", () => {
  test("should be rendered", () => {
    const { asFragment } = render(<PersonForm />);
    expect(asFragment()).toMatchSnapshot();
  });
  test("name: string!", () => {
    const { getByTestId } = render(<PersonForm />);
    const nameField = getByTestId("personName");
    expect(nameField).toBeInTheDocument();
  });
  test("should disable submit when name is empty", async () => {
    const { container, findByText, getByPlaceholderText } = render(
      <PersonForm person={{ personName: "Marco" }} />
    );
    const nameInput = getByPlaceholderText("Name");
    const submitButton = container.querySelector("button[type=submit]");

    expect(submitButton).toBeDisabled();

    await fireEvent.change(nameInput, { target: { value: "" } });
    await fireEvent.blur(nameInput);

    expect(await findByText(/required/i)).not.toBeNull();
    expect(submitButton).toBeDisabled();

    await fireEvent.change(nameInput, {
      target: { name: "personName", value: "Samuele" },
    });

    expect(await findByText(/required/i)).not.toBeNull();
    expect(submitButton).toBeEnabled();
  });
});
