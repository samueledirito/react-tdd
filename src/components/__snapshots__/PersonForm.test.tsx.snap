// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PersonForm should be rendered 1`] = `
<DocumentFragment>
  <form
    action="#"
  >
    <div
      class="MuiFormControl-root MuiTextField-root"
      data-testid="personName"
    >
      <label
        class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-filled"
        data-shrink="false"
      >
        Name
      </label>
      <div
        class="MuiInputBase-root MuiFilledInput-root MuiFilledInput-underline MuiInputBase-formControl"
      >
        <input
          aria-invalid="false"
          class="MuiInputBase-input MuiFilledInput-input"
          name="personName"
          placeholder="Name"
          type="text"
          value=""
        />
      </div>
      <p
        class="MuiFormHelperText-root MuiFormHelperText-contained"
      >
        Name of this person
      </p>
    </div>
    <button
      class="MuiButtonBase-root MuiButton-root MuiButton-text Mui-disabled Mui-disabled"
      disabled=""
      tabindex="-1"
      type="submit"
    >
      <span
        class="MuiButton-label"
      >
        Save
      </span>
    </button>
  </form>
</DocumentFragment>
`;
